### ISSUE TITLE
All salt-minions died with an unhandled exceptions while salt-master was unavailable due to maintenance

### ISSUE THREAD

**Issue by Whissi** [saltstack/salt#20735]
Due to a maintenance window (3h) which started at 18.00 the server running *salt-master* restarted several times and lost/restored network connection.

After we finished maintenance we noticed that **every** salt-minion in our network died at the same time (~120 minions) with an unhandled exception:
```
2015-02-16 18:41:30,335 [salt.utils.schedule                         ][ERROR   ] Unhandled exception running mine.update
Traceback (most recent call last):
  File "/usr/lib64/python2.7/site-packages/salt/utils/schedule.py", line 437, in handle_func
    ret['return'] = self.functions[func]()
  File "/usr/lib64/python2.7/site-packages/salt/modules/mine.py", line 99, in update
    load['tok'] = _auth().gen_token('salt')
  File "/usr/lib64/python2.7/site-packages/salt/modules/mine.py", line 25, in _auth
    __context__['auth'] = salt.crypt.SAuth(__opts__)
  File "/usr/lib64/python2.7/site-packages/salt/crypt.py", line 810, in __init__
    self.crypticle = self.__authenticate()
  File "/usr/lib64/python2.7/site-packages/salt/crypt.py", line 831, in __authenticate
    self.opts.get('auth_tries', 1)
  File "/usr/lib64/python2.7/site-packages/salt/crypt.py", line 649, in sign_in
    raise SaltClientError('Attempt to authenticate with the salt master failed')
SaltClientError: Attempt to authenticate with the salt master failed
2015-02-16 18:51:10,657 [salt                                        ][ERROR   ] Attempt to authenticate with the salt master failed
```

```
# salt --versions-report
           Salt: 2014.7.1
         Python: 2.7.9 (default, Feb  8 2015, 01:34:33)
         Jinja2: 2.7.3
       M2Crypto: 0.22
 msgpack-python: 0.4.5
   msgpack-pure: Not Installed
       pycrypto: 2.6.1
        libnacl: Not Installed
         PyYAML: 3.11
          ioflo: Not Installed
          PyZMQ: 14.5.0
           RAET: Not Installed
            ZMQ: 4.0.5
           Mako: Not Installed
```

**Comment by cachedout** [comment]
The mine needs to gracefully handle auth failures. That should be easy to fix.

**Comment by cachedout** [comment]
This issue needs a test case written, please.

**Comment by rallytime** [comment]
So we use the "needs testcase" on open and closed issues as well as pull requests. Once the tests have been written, we link them and remove the label (so we can help keep track of this that are fixed and need tests vs. things that need tests on their own vs. things that needs tests and are still bugs). As such, I think this particular issue is fixed with the PR above. @Whissi Are you in a position to give that a try?

**Comment by Whissi** [comment]
@rallytime: I couldn't test the PR because it doesn't apply against salt-2014.7.1.

However I tested against git (2015.2.0-742-g58faaa5) and it seems to work. While the git version now logs
```
2015-03-16 00:53:00,798 [salt.minion                              ][WARNING ][7531] The minion failed to return the job information for job 20150316005000550734. This is often due to the master being shut down or overloaded. If the master is running consider increasing the worker_threads value.
```
a minion still running *salt-2014.7.1* had thrown the known exception:
```
2015-03-16 00:49:33,218 [salt.utils.schedule                         ][ERROR   ] Unhandled exception running mine.update
Traceback (most recent call last):
  File "/usr/lib64/python2.7/site-packages/salt/utils/schedule.py", line 437, in handle_func
    ret['return'] = self.functions[func]()
  File "/usr/lib64/python2.7/site-packages/salt/modules/mine.py", line 99, in update
    load['tok'] = _auth().gen_token('salt')
  File "/usr/lib64/python2.7/site-packages/salt/modules/mine.py", line 25, in _auth
    __context__['auth'] = salt.crypt.SAuth(__opts__)
  File "/usr/lib64/python2.7/site-packages/salt/crypt.py", line 810, in __init__
    self.crypticle = self.__authenticate()
  File "/usr/lib64/python2.7/site-packages/salt/crypt.py", line 831, in __authenticate
    self.opts.get('auth_tries', 1)
  File "/usr/lib64/python2.7/site-packages/salt/crypt.py", line 649, in sign_in
    raise SaltClientError('Attempt to authenticate with the salt master failed')
SaltClientError: Attempt to authenticate with the salt master failed
```
So this looks fixed for me. Thanks!



<hr>
**How did I reproduce the problem?**
To simulate the problem I added a iptables rule on master to drop traffic on TCP port 4505-4506 for the two test minions.

**Comment by rallytime** [comment]
@Whissi Thanks for checking that out and confirming the fix on the 2015.2 branch. That change has not been applied to the 2014.7 branch as the fix took place on re-factored code. Since you've reported that this is fixed, I am going to close this issue. If you still see problems, leave a comment and we can re-open the issue.
